sbin/tunefs/tunefs.8
.\"	$OpenBSD: tunefs.8,v 1.26 2010/10/18 15:01:40 jmc Exp $
.\"	$NetBSD: tunefs.8,v 1.36 2004/12/20 10:28:47 hubertf Exp $
.\"
.\" Copyright (c) 1983, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)tunefs.8	8.3 (Berkeley) 5/3/95
.\"
.Dd $Mdocdate: October 18 2010 $
.Dt TUNEFS 8
.Os
.Sh NAME
.Nm tunefs
.Nd tune up an existing file system
.Sh SYNOPSIS
.Nm
.Bk -words
.Op Fl AFN
.Op Fl e Ar maxbpg
.Op Fl g Ar avgfilesize
.Op Fl h Ar avgfpdir
.Op Fl l Ar logsize
.Op Fl m Ar minfree
.\" .Op Fl n Ar soft_dependency_enabling
.Op Fl o Ar optimize_preference
.Ar special | filesys
.Ek
.Sh DESCRIPTION
.Nm
is designed to change the dynamic parameters of a file system
which affect the layout policies.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl A
Cause the values to be updated in all the alternate
superblocks instead of just the standard superblock.
If this option is not used,
then use of a backup superblock by
.Xr fsck 8
will lose anything changed by
.Nm .
.Fl A
is ignored when
.Fl N
is specified.
.It Fl e Ar maxbpg
This indicates the maximum number of blocks any single file can
allocate out of a cylinder group before it is forced to begin
allocating blocks from another cylinder group.
Typically this value is set to about one quarter of the total blocks
in a cylinder group.
The intent is to prevent any single file from using up all the
blocks in a single cylinder group,
thus degrading access times for all files subsequently allocated
in that cylinder group.
The effect of this limit is to cause big files to do long seeks
more frequently than if they were allowed to allocate all the blocks
in a cylinder group before seeking elsewhere.
For file systems with exclusively large files,
this parameter should be set higher.
.It Fl F
Indicates that
.Ar special
is a file system image, rather than a device name or file system mount point.
.Ar special
will be accessed
.Sq as-is .
.It Fl g Ar avgfilesize
This specifies the expected average file size.
.It Fl h Ar avgfpdir
This specifies the expected number of files per directory.
.It Fl l Ar logsize
This value specifies the size of the in-filesystem journaling log file.
The default journaling log file size is described in
.Xr wapbl 4 .
Specifying a size of zero will cause the in-filesystem journaling log file
to be removed the next time the filesystem is mounted.
The size of an existing in-filesystem journaling log file cannot be changed
directly.
You need to first set the log file size to zero, then mount the
filesystem without logging enabled (which will remove the log without
creating a new one), unmount, set the size to the new value and finally
re-mount with logging enabled.
.It Fl m Ar minfree
This value specifies the percentage of space held back
from normal users; the minimum free space threshold.
The default value is set during creation of the filesystem; see
.Xr newfs 8 .
This value can be set to zero, however up to a factor of three
in throughput will be lost over the performance obtained at a 5%
threshold.
Note that if the value is raised above the current usage level,
users will be unable to allocate files until enough files have
been deleted to get under the higher threshold.
.It Fl N
Display all the settable options
(after any changes from the tuning options)
but do not cause any of them to be changed.
.It Fl o Ar optimize_preference
The file system can either try to minimize the time spent
allocating blocks, or it can attempt to minimize the space
fragmentation on the disk.
If the value of
.Ar minfree
(see above) is less than 5%,
then the file system should optimize for space to avoid
running out of full sized blocks.
For values of
.Ar minfree
greater than or equal to 5%,
fragmentation is unlikely to be problematical, and
the file system can be optimized for time.
.Pp
.Ar optimize_preference
can be specified as either
.Li space
or
.Li time .
.El
.Sh SEE ALSO
.Xr wapbl 4 ,
.Xr fs 5 ,
.Xr dumpfs 8 ,
.Xr fsck_ffs 8 ,
.Xr growfs 8 ,
.Xr newfs 8
.Rs
.%A M. McKusick
.%A W. Joy
.%A S. Leffler
.%A R. Fabry
.%T "A Fast File System for UNIX"
.%J "ACM Transactions on Computer Systems 2"
.%N 3
.%P pp 181-197
.%D August 1984
.Re
.Sh HISTORY
The
.Nm
command appeared in
.Bx 4.2 .
.Sh BUGS
This program should work on mounted and active file systems.
Because the super-block is not kept in the buffer cache,
the changes will only take effect if the program
is run on unmounted file systems.
To change the root file system, the system must be rebooted
after the file system is tuned.
.Pp
.\" Take this out and a Unix Demon will dog your steps from now until
.\" the time_t's wrap around.
You can tune a file system, but you can't tune a fish.


sbin/tunefs/tunefs.c
/*	$OpenBSD: tunefs.c,v 1.32 2013/06/11 16:42:05 deraadt Exp $	*/
/*	$NetBSD: tunefs.c,v 1.33 2005/01/19 20:46:16 xtraeme Exp $	*/

/*
 * Copyright (c) 1983, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/*
 * tunefs: change layout parameters to an existing file system.
 */
#include <sys/param.h>

#include <ufs/ufs/dinode.h>
#include <ufs/ffs/fs.h>
#include <ufs/ufs/ufs_wapbl.h>
#include <ufs/ffs/ffs_extern.h>

#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <fstab.h>
#include <paths.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <util.h>

/* the optimization warning string template */
#define	OPTWARN	"should optimize for %s with minfree %s %d%%"

union {
	struct	fs sb;
	char pad[MAXBSIZE];
} sbun;
#define	sblock sbun.sb
char buf[MAXBSIZE];

int	fi;
long	dev_bsize = 512;
int	is_ufs2 = 0;
off_t	sblockloc;

static off_t sblock_try[] = SBLOCKSEARCH;

static	void	bwrite(daddr_t, char *, int, const char *);
static	void	bread(daddr_t, char *, int, const char *);
static	void	change_log_info(long long);
static	int	getnum(const char *, const char *, int, int);
static	void	getsb(struct fs *, const char *);
static	int	openpartition(char *, int, char **);
static	void	show_log_info(void);
static	void	usage(void);

int
main(int argc, char *argv[])
{
#define	OPTSTRING	"AFNe:g:h:l:m:o:"
	int		i, ch, Aflag, Fflag, Nflag, openflags;
	char		*special;
	const char	*chg[2];
	int		maxbpg, minfree, optim;
	int		avgfilesize, avgfpdir;
	long long	logfilesize;

	Aflag = Fflag = Nflag = 0;
	maxbpg = minfree = optim = -1;
	avgfilesize = avgfpdir = -1;
	logfilesize = -1;
	chg[FS_OPTSPACE] = "space";
	chg[FS_OPTTIME] = "time";

	while ((ch = getopt(argc, argv, OPTSTRING)) != -1) {
		switch (ch) {

		case 'A':
			Aflag++;
			break;

		case 'F':
			Fflag++;
			break;

		case 'N':
			Nflag++;
			break;

		case 'e':
			maxbpg = getnum(optarg,
			    "maximum blocks per file in a cylinder group",
			    1, INT_MAX);
			break;

		case 'g':
			avgfilesize = getnum(optarg,
			    "average file size", 1, INT_MAX);
			break;

		case 'h':
			avgfpdir = getnum(optarg,
			    "expected number of files per directory",
			    1, INT_MAX);
			break;

		case 'l':
			if (scan_scaled(optarg, &logfilesize) == -1 ||
			    logfilesize < 0)
				errx(1, "invalid number `%s' for journal log "
				    "file size", optarg);
			break;

		case 'm':
			minfree = getnum(optarg,
			    "minimum percentage of free space", 0, 99);
			break;

		case 'o':
			if (strcmp(optarg, chg[FS_OPTSPACE]) == 0)
				optim = FS_OPTSPACE;
			else if (strcmp(optarg, chg[FS_OPTTIME]) == 0)
				optim = FS_OPTTIME;
			else
				errx(10,
				    "bad %s (options are `space' or `time')",
				    "optimization preference");
			break;

		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind; 
	if (argc != 1)
		usage();

	special = argv[0];
	openflags = Nflag ? O_RDONLY : O_RDWR;
	if (Fflag)
		fi = open(special, openflags);
	else
		fi = openpartition(special, openflags, &special);
	if (fi == -1)
		err(1, "%s", special);
	getsb(&sblock, special);

#define CHANGEVAL(old, new, type, suffix) do				\
	if ((new) != -1) {						\
		if ((new) == (old))					\
			warnx("%s remains unchanged at %d%s",		\
			    (type), (old), (suffix));			\
		else {							\
			warnx("%s changes from %d%s to %d%s",		\
			    (type), (old), (suffix), (new), (suffix));	\
			(old) = (new);					\
		}							\
	} while (/* CONSTCOND */0)

	warnx("tuning %s", special);
	CHANGEVAL(sblock.fs_maxbpg, maxbpg,
	    "maximum blocks per file in a cylinder group", "");
	CHANGEVAL(sblock.fs_minfree, minfree,
	    "minimum percentage of free space", "%");
	if (minfree != -1) {
		if (minfree >= MINFREE &&
		    sblock.fs_optim == FS_OPTSPACE)
			warnx(OPTWARN, "time", ">=", MINFREE);
		if (minfree < MINFREE &&
		    sblock.fs_optim == FS_OPTTIME)
			warnx(OPTWARN, "space", "<", MINFREE);
	}
	if (optim != -1) {
		if (sblock.fs_optim == optim) {
			warnx("%s remains unchanged as %s",
			    "optimization preference",
			    chg[optim]);
		} else {
			warnx("%s changes from %s to %s",
			    "optimization preference",
			    chg[sblock.fs_optim], chg[optim]);
			sblock.fs_optim = optim;
			if (sblock.fs_minfree >= MINFREE &&
			    optim == FS_OPTSPACE)
				warnx(OPTWARN, "time", ">=", MINFREE);
			if (sblock.fs_minfree < MINFREE &&
			    optim == FS_OPTTIME)
				warnx(OPTWARN, "space", "<", MINFREE);
		}
	}
	CHANGEVAL(sblock.fs_avgfilesize, avgfilesize,
	    "average file size", "");
	CHANGEVAL(sblock.fs_avgfpdir, avgfpdir,
	    "expected number of files per directory", "");

	if (logfilesize >= 0)
		change_log_info(logfilesize);

	if (Nflag) {
		fprintf(stdout, "tunefs: current settings of %s\n", special);
		fprintf(stdout, "\tmaximum contiguous block count %d\n",
		    sblock.fs_maxcontig);
		fprintf(stdout,
		    "\tmaximum blocks per file in a cylinder group %d\n",
		    sblock.fs_maxbpg);
		fprintf(stdout, "\tminimum percentage of free space %d%%\n",
		    sblock.fs_minfree);
		fprintf(stdout, "\toptimization preference: %s\n",
		    chg[sblock.fs_optim]);
		fprintf(stdout, "\taverage file size: %d\n",
		    sblock.fs_avgfilesize);
		fprintf(stdout,
		    "\texpected number of files per directory: %d\n",
		    sblock.fs_avgfpdir);
		show_log_info();
		fprintf(stdout, "tunefs: no changes made\n");
		exit(0);
	}

	memcpy(buf, (char *)&sblock, SBLOCKSIZE);
	bwrite(sblockloc, buf, SBLOCKSIZE, special);
	if (Aflag)
		for (i = 0; i < sblock.fs_ncg; i++)
			bwrite(fsbtodb(&sblock, cgsblock(&sblock, i)),
			    buf, SBLOCKSIZE, special);
	close(fi);
	exit(0);
}

static void
show_log_info(void)
{
	const char *loc;
	uint64_t size, blksize, logsize;
	int print;

	switch (sblock.fs_journal_location) {
	case UFS_WAPBL_JOURNALLOC_NONE:
		print = blksize = 0;
		/* nothing */
		break;
	case UFS_WAPBL_JOURNALLOC_END_PARTITION:
		loc = "end of partition";
		size = sblock.fs_journallocs[UFS_WAPBL_EPART_COUNT];
		blksize = sblock.fs_journallocs[UFS_WAPBL_EPART_BLKSZ];
		print = 1;
		break;
	case UFS_WAPBL_JOURNALLOC_IN_FILESYSTEM:
		loc = "in filesystem";
		size = sblock.fs_journallocs[UFS_WAPBL_INFS_COUNT];
		blksize = sblock.fs_journallocs[UFS_WAPBL_INFS_BLKSZ];
		print = 1;
		break;
	default:
		loc = "unknown";
		size = blksize = 0;
		print = 1;
		break;
	}

	if (print) {
		logsize = size * blksize;

		printf("\tjournal log file location: %s\n", loc);
		printf("\tjournal log file size: ");
		if (logsize == 0)
			printf("0");
		else {
			char sizebuf[FMT_SCALED_STRSIZE+1];
			if (fmt_scaled(logsize, sizebuf) == 0) {
				sizebuf[FMT_SCALED_STRSIZE] = '\0';
				printf("%s (%lld bytes)", sizebuf, logsize);
			} else
				printf("%lld bytes", logsize);
		}
		printf("\n");
		printf("\tjournal log flags:");
		if (sblock.fs_journal_flags & UFS_WAPBL_FLAGS_CREATE_LOG)
			printf(" create-log");
		if (sblock.fs_journal_flags & UFS_WAPBL_FLAGS_CLEAR_LOG)
			printf(" clear-log");
		printf("\n");
	}
}

static void
change_log_info(long long logfilesize)
{
	/*
	 * NOTES:
	 *  - only operate on in-filesystem log sizes
	 *  - can't change size of existing log
	 *  - if current is same, no action
	 *  - if current is zero and new is non-zero, set flag to create log
	 *    on next mount
	 *  - if current is non-zero and new is zero, set flag to clear log
	 *    on next mount
	 */
	int in_fs_log;
	uint64_t old_size;

	old_size = 0;
	switch (sblock.fs_journal_location) {
	case UFS_WAPBL_JOURNALLOC_END_PARTITION:
		in_fs_log = 0;
		old_size = sblock.fs_journallocs[UFS_WAPBL_EPART_COUNT] *
		    sblock.fs_journallocs[UFS_WAPBL_EPART_BLKSZ];
		break;

	case UFS_WAPBL_JOURNALLOC_IN_FILESYSTEM:
		in_fs_log = 1;
		old_size = sblock.fs_journallocs[UFS_WAPBL_INFS_COUNT] *
		    sblock.fs_journallocs[UFS_WAPBL_INFS_BLKSZ];
		break;

	case UFS_WAPBL_JOURNALLOC_NONE:
	default:
		in_fs_log = 0;
		old_size = 0;
		break;
	}

	if (logfilesize == 0) {
		/*
		 * Don't clear out the locators - the kernel might need
		 * these to find the log!  Just set the "clear the log"
		 * flag and let the kernel do the rest.
		 */
		sblock.fs_journal_flags |= UFS_WAPBL_FLAGS_CLEAR_LOG;
		sblock.fs_journal_flags &= ~UFS_WAPBL_FLAGS_CREATE_LOG;
		warnx("log file size cleared from %llu", old_size);
		return;
	}

	if (!in_fs_log && logfilesize > 0 && old_size > 0)
		errx(1, "can't change size of non-in-filesystem log");

	if (old_size == (uint64_t)logfilesize && logfilesize > 0) {
		/* no action */
		warnx("log file size remains unchanged at %lld", logfilesize);
		return;
	}

	if (old_size == 0) {
		/* create new log of desired size next mount */
		sblock.fs_journal_location = UFS_WAPBL_JOURNALLOC_IN_FILESYSTEM;
		sblock.fs_journallocs[UFS_WAPBL_INFS_ADDR] = 0;
		sblock.fs_journallocs[UFS_WAPBL_INFS_COUNT] = logfilesize;
		sblock.fs_journallocs[UFS_WAPBL_INFS_BLKSZ] = 0;
		sblock.fs_journallocs[UFS_WAPBL_INFS_INO] = 0;
		sblock.fs_journal_flags |= UFS_WAPBL_FLAGS_CREATE_LOG;
		sblock.fs_journal_flags &= ~UFS_WAPBL_FLAGS_CLEAR_LOG;
		warnx("log file size set to %lld", logfilesize);
	} else {
		errx(1,
		    "can't change existing log size from %llu to %lld",
		     old_size, logfilesize);
	} 
}

static int
getnum(const char *num, const char *desc, int min, int max)
{
	int		n;
	const char	*errstr;

	n = strtonum(num, min, max, &errstr);
	if (errstr != NULL)
		errx(1, "Invalid number `%s' for %s: %s", num, desc, errstr);
	return (n);
}

static void
usage(void)
{
	extern char *__progname;

	fprintf(stderr,
	    "usage: %s [-AFN] [-e maxbpg] [-g avgfilesize] "
	    "[-h avgfpdir] [-m minfree]\n"
	    "\t[-o optimize_preference] special | filesys\n",
	    __progname);

	exit(2);
}

static void
getsb(struct fs *fs, const char *file)
{
	int i;

	for (i = 0; ; i++) {
		if (sblock_try[i] == -1)
			errx(5, "cannot find filesystem superblock");
		bread(sblock_try[i] / dev_bsize, (char *)fs, SBLOCKSIZE, file);
		switch(fs->fs_magic) {
		case FS_UFS2_MAGIC:
			is_ufs2 = 1;
			/*FALLTHROUGH*/
		case FS_UFS1_MAGIC:
			break;
		default:
			continue;
		}
		if (!is_ufs2 && sblock_try[i] == SBLOCK_UFS2)
			continue;
		if ((is_ufs2 || fs->fs_flags & FS_FLAGS_UPDATED)
		    && fs->fs_sblockloc != sblock_try[i])
			continue;
		break;
	}

	dev_bsize = fs->fs_fsize / fsbtodb(fs, 1);
	sblockloc = sblock_try[i] / dev_bsize;
}

static void
bwrite(daddr_t blk, char *buffer, int size, const char *file)
{
	off_t	offset;

	offset = (off_t)blk * dev_bsize;
	if (lseek(fi, offset, SEEK_SET) == -1)
		err(6, "%s: seeking to %lld", file, (long long)offset);
	if (write(fi, buffer, size) != size)
		err(7, "%s: writing %d bytes", file, size);
}

static void
bread(daddr_t blk, char *buffer, int cnt, const char *file)
{
	off_t	offset;
	int	i;

	offset = (off_t)blk * dev_bsize;
	if (lseek(fi, offset, SEEK_SET) == -1)
		err(4, "%s: seeking to %lld", file, (long long)offset);
	if ((i = read(fi, buffer, cnt)) != cnt)
		errx(5, "%s: short read", file);
}

static int
openpartition(char *name, int flags, char **devicep)
{
	char		rawspec[MAXPATHLEN], *p;
	struct fstab	*fs;
	int		fd;

	fs = getfsfile(name);
	if (fs) {
		if ((p = strrchr(fs->fs_spec, '/')) != NULL) {
			snprintf(rawspec, sizeof(rawspec), "%.*s/r%s",
			    (int)(p - fs->fs_spec), fs->fs_spec, p + 1);
			name = rawspec;
		} else
			name = fs->fs_spec;
	}
	fd = opendev(name, flags, 0, devicep);
	if (fd == -1 && errno == ENOENT)
		devicep = &name;
	return (fd);
}
